west/sign: Rework of how MCUboot partitions are taken from DT

de-nordic opened the pull request:
The commit moves from using chosen `zephyr,flash` to flash device where chosen `zephyr,code-partition` is placed. There has been special class created MCUbootDTConfig that is helper class for accessing MCUboot type partitions. The class incorporates edt_flash_node logic, and the method has been removed.

Object of the MCUbootDTConfig class is now used by mcuboot_swap_validate method, which replaces edt_flash_params, that is supposed to validate partition layout for swap configuration. mcuboot_swap_validate extends edt_flash_params with additional checks and warning logs.
The mcuboot_swap_validate maintains compatibility with legacy behaviour where single slot MCUboot configuration is assumed if slot1_partition is not found.

**Note**: This is base work for separating current validation into swap-move and single application mode, and addition of DirectXIP and swap-scratch.

- [x] Documentation is missing
- [x] Commit with binary file validation against image slot size
- [x]  Separate swap-move and single application mode
- [x]  Support for swap-scratch
- [x]  Support for DirectXIP
- [x] Release notes 

**CONTENTS**
The PR consists of commits that:
 1) rewrite swap/single slot validation
 2) add application size validation against slot size
 3) split swap/single validation to separate swap and separate single slot validation
 4) add esp32 overlay required 
 5) Kconfig options to let application know what mode of operation has been set for MCUboot
 6) swap-scratch validation support

The sing script commits can be squashed, they are separated to make it easier to understand steps taken.

Fixes #54373.

de-nordic committed:
west/sign: Rework of how MCUboot partitions are taken from DT

The commit moves from using chosen `zephyr,flash` to flash
device where chosen `zephyr,code-partition` is placed.
There has been special class created MCUbootDTConfig that
is helper class for accessing MCUboot type partitions.
The class incorporates edt_flash_node logic, and the method
has been removed.

Object of the MCUbootDTConfig class is now used by
mcuboot_swap_validate method, which replaces edt_flash_params,
that is supposed to validate partition layout for swap
configuration. mcuboot_swap_validate extends edt_flash_params
with additional checks and warning logs.
The mcuboot_swap_validate maintains compatibility with
legacy behaviour where single slot MCUboot configuration is
assumed if slot1_partition is not found.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mbolivar-nordic commented:
@de-nordic the pull request description is doing a great job of saying what is happening, but not why. Why is this change needed?

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -205,6 +205,142 @@ def sign(self, command, build_dir, build_conf, formats):
+            if op_size != cp_size:
+                log.wrn(f"slot0_partition and slot1_partition differ in size:"
+                             " ${cp_size} vs ${op_size}. Smaller size will be used.")
+            if op_size < cp_size:
+                cp_size = op_size
```
Indent one extra

de-nordic commented:
> @de-nordic the pull request description is doing a great job of saying what is happening, but not why. Why is this change needed?

The MCUbootDTConfig  has been provided to make it easier to get MCUboot related partition configuration, to use with series of `mcuboot_swap_validate`  type methods: `mcuboot_single_validate`, `mcuboot_scratch_validate` and `mcuboot_directxip_validate` . These will validate build app vs DT partitions, assuming MCUboot configuration, and will either refuse singing or return proper parameters for signing, eventually giving some warnings (for example single configuration will warn that there are separate slot0 and slot1, which probably wastes space).
Above functions are missing now, but to be able to reuse parts of edt_ methods code, I have decoded to move them to MCUbootDTConfig  class.

mbolivar-nordic commented:
@de-nordic sorry I was unclear -- why is the change of chosen node needed? Won't this break existing users?

de-nordic commented:
> @de-nordic sorry I was unclear -- why is the change of chosen node needed? Won't this break existing users?

No it should not.
While signing image we expect it to be selected by `zephyr,code-partition`, so we should also figure out flash the `zephyr,code-partition` is placed on, instead of using `zephyr,flash`.
If user has previously set `zephyr,code-partition` to partition placed on something different than `zephyr,flash` than it would probably not work for the user anyway. So user would end up correcting `zephyr,flash` to point to flash where `zephyr,code-partition` is placed, or moving `zephyr,code-partition` to that flash.
Basically in previous code we were a little bit out-of-sync with DTS: we would expect `zephyr,code-partition` to point to binary we start with, but use `zephyr,flash` to sign image (which have to be the same for this to work anyway). Common stuff we check was slot labels.

mbolivar-nordic reviewed (dismissed):
The explanation makes sense to me. Thank you.

Reassigning the PR since @nordicjm has started taking overall responsibility for flashing, but it LGTM.

de-nordic commented:
> The explanation makes sense to me. Thank you.
> 
> Reassigning the PR since @nordicjm has started taking overall responsibility for flashing, but it LGTM.

Thanks.

Anyway, it seems to break esp32 which I need to take look into. It seems that they select CONFIG_USE_DT_CODE_PARTITION, vaie CONFIG_BOOTLOADER_MCUBOOT, even though they do not use th zephyr,code-partition at all. This is misleading.
I also found that they use sometimes `zephyr,flash` but also seem to hardcode, in Kconfigs, specific flash device via DT path.

de-nordic committed:
samples/boards/esp32: Add chosen zpehyr,code-partition to overlay

This is used by west sign to figure out from which slot and which
partition application is run from.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Separate single slot and swap vaidation for MCUboot

Separates single slot validation of MCUboot configuration
into separate method, from swap algorithm validation.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Add application size check agains slot size

The commit adds check for swap-move algorithm where application
size, expanded by size off required additions, is compared agains
available slot size.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Add Kconfig options for MCUboot mode of operation

Commits adds Kconfig options that inform application on
MCUboot application update algorithm selected and modifies
west sign script to use new Kconfig while validating image
and DT configuration for signing.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Add MCUboot swap-scratch validation

Adds validation method for MCUboot swap-scratch algorithm.
Commit also adds CONFIG_MCUBOOT_BOOTLOADER_USES_SWAP_SCRATCH
Kconfig option that allows to indicate that MCUboot has been
configured for swap-scratch mode of operation.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Add MCUboot DirectXIP validation

Added validating application image for DirectXIP execution.
The commits adds CONFIG_MCUBOOT_BOOTLOADER_USES_DIRECT_XIP Kconfig
option that allows to indicate that application is built with
MCUboot in DirectXIP mode.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
west/sign: Add option to encode application slot address to header

Commits adds CONFIG_MCUBOOT_IMGTOOL_ENCODE_SLOT_ADDRESS option
that allows to encode application slot address into header.
Supporting code has been added to west sign.
This is useful for doing basic check, during DFU, whether applicaiton
has been build for slot it is going to boot from.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
docs/release/3.4: Add notes on changes to west sign

The commit adds release notes regarding changes in west sign
and addition of Kconfig options:
 CONFIG_MCUBOOT_BOOTLOADER_USES_SWAP
 CONFIG_MCUBOOT_BOOTLOADER_USES_SINGLE
 CONFIG_MCUBOOT_BOOTLOADER_USES_SWAP_SCRATCH
 CONFIG_MCUBOOT_BOOTLOADER_USES_DIRECT_XIP
 CONFIG_MCUBOOT_IMGTOOL_ENCODE_SLOT_ADDRESS

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on samples/boards/esp32/flash_encryption/boards/esp32.overlay:
```diff
@@ -7,3 +7,9 @@
 &flash0 {
```
typo in this commit: `samples/boards/esp32: Add chosen zpehyr,code-partition to overlay` *zephyr

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
+            _, op_size = other.get_parameters()
+            # Always go with smaller partition size, as swap will not be possible
+            if op_size != cp_size:
+                log.wrn("slot0_partition and slot1_partition differ in size:"
+                        f" ${cp_size} vs ${op_size}. Smaller size will be used.")
```
Have seen some people use different slot sizes because of the trailer information stored on one slot only, might be an idea (in future) to check which slot is larger and ignore the error if it's that one

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
+                if op_size < cp_size:
+                    cp_size = op_size
+        else:
+            log.wrn('slot1_partition not found, assuming single slot configuration')
```
I think this might become a bit verbose to see this every time if the user is using single slot mode, debug level log might be better suited

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -12,6 +12,7 @@
 import shlex
```
typo in commit: size, expanded by size off required additions, is compared agains
*against

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -422,6 +423,20 @@ def get_code_partition(self):
+    def mcuboot_magic_size(flash_alignment=8):
+        '''Get MCUboot magic size
+
+        Magic size is is 16 octets aligned to flash write alignment (min. 8) + 3 flags,
```
-is

nordicjm reviewed (changes requested):
(not fully reviewed)

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -490,11 +505,44 @@ def mcuboot_swap_validate(mcbdtconf, bin_size, quiet=False):
+
+            bin_size_wm = bin_size + __class__.mcuboot_magic_size(write_alignment)
+
+            # Binary size can not be greater than slot size without last page,
+            # as the last page is used for move algorithm.
+            if bin_size_wm > (cp_size - erase_size):
+                log.die("Application image with added MCUboot magic will not fit"
+                        " in application slot, for swap-move configured MCUboot.\n"
+                        f"Binary size is {bin_size}, with magic it is {bin_size_wm}.\n"
+                        f"Max allowed slot size is {cp_size} - {erase_size}")
```
Didn't see that a Kconfig has been added (in a later commit), this selection probably needs updating depending upon if a scratch partition is present, and be linked with sysbuild. For example, most stm32 boards should be using swap using scratch by default, they cannot operate in swap using move mode

nordicjm replied:
Ping @tejlmand for note on sysbuild integration

de-nordic replied:
This part is for swap-move only, scratch validation is done by separate function.

> For example, most stm32 boards should be using swap using scratch by default, they cannot operate in swap using move mode

OK, so defconfigs for these boards would need to be altered.

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -520,6 +520,81 @@ def mcuboot_swap_validate(mcbdtconf, bin_size, quiet=False):
+                    " slot0_partition. Currently selected node is labeled"
+                    " " + str(code.labels))
+
+        if 'slot1_partition' and 'scratch_partition' in code.labels:
+            log.die("chosen zephyr,code-partition may not be slot0_partition,"
+                    " slot1_partition and scratch_partition at the same time")
```
This cannot happen

de-nordic replied:
Unless somebody does two labels to the same node. I can remove these checks, if not wanted.

nordicjm commented on scripts/west_commands/sign.py:
```diff
@@ -652,6 +652,83 @@ def mcuboot_single_validate(mcbdtconf, bin_size, quiet=False):
+                    "MCUboot is not yet able to DirectXIP from external flash")
+
+        if mcbdtconf.get_partition('scratch_partition'):
+            log.wrn("scratch_partition not needed for MCUboot in single applicaiton"
+                    " mode")
```
Wrong comment

de-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -237,6 +237,264 @@ def sign(self, command, build_dir, build_conf, formats):
+            partition = ImgtoolSigner.DTPartition(node)
+
+            if not partition.is_internal():
+                log.die("west sign currently does not support signing image"
+                        " built for code-partition that has not been placed"
+                        " on internal flash")
```
The board is complete, logic for checking whether partition is internal is broken, as it does not assume that there might be more than one compatible assigned to flash node.
